### PR TITLE
implemented 'nimble publish'

### PR DESCRIPTION
**Pull request by Araq**

### CODE REVIEW

File: src/nimblepkg/version.nim [hunk 1]
```diff
@@ -29,6 +29,9 @@ type
     of verAny:
       nil
 
+  ## Tuple containing package name and version range.
+  PkgTuple* = tuple[name: string, ver: VersionRange]
```
Comment (dom96) [inline src/nimblepkg/version.nim]: Why did you move this here?

File: src/nimblepkg/publish.nim [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+proc createFork(a: Auth) =
+  discard postContent("https://api.github.com/repos/nim-lang/packages/forks",
+      extraHeaders=("Authorization: Basic $1\c\L" % a.token) &
+        "Content-Type: application/x-www-form-urlencoded\c\L" &
+        "Accept: */*\c\L")
+
+proc createPullRequest(a: Auth; packageName: string) =
+  echo "creating PR"
+  discard postContent("https://api.github.com/repos/nim-lang/packages/pulls",
```
Comment (dom96) [inline src/nimblepkg/publish.nim]: Check what the result is here.

File: src/nimblepkg/publish.nim [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+      extraHeaders=("Authorization: Basic $1\c\L" % a.token) &
```
Comment (dom96) [inline src/nimblepkg/publish.nim]: I doubt this method of auth will work if I have got two-factor authentication enabled. Could we not do oauth?

File: src/nimblepkg/publish.nim (hunk 2 above)
Comment (dom96) [inline src/nimblepkg/publish.nim]: Also, we're back to depending on OpenSSL :(

File: src/nimblepkg/publish.nim [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+proc getPackageOriginUrl(a: Auth): string =
+  ## Adds 'user:pw' to the URL so that the user is not asked *again* for it.
+  ## We need this for 'git push'.
+  let (output, exitCode) = doCmdEx("git config --get remote.origin.url")
+  result = "origin"
+  if exitCode == 0:
+    result = output.string.strip
+    if result.endsWith(".git"): result.setLen(result.len - 4)
+    if result.startsWith("https://"):
+      result = "https://" & a.user & ':' & a.pw & '@' &
```
Comment (dom96) [inline src/nimblepkg/publish.nim]: Use the URI module here.

File: src/nimblepkg/publish.nim [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+          result["https://".len .. ^1]
+
+proc publish*(p: PackageInfo) =
+  ## Publishes the package p.
+  let auth = getGithubAuth()
+  let parent = os.getCurrentDir().parentDir()
```
Comment (dom96) [inline src/nimblepkg/publish.nim]: This should probably use getTempDir.

File: src/nimblepkg/publish.nim [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+  var pkgsDir = parent / "nimble-packages-fork"
+  if not forkExists(auth):
+    createFork(auth)
+    echo "waiting 10s to let Github create a fork ..."
```
Comment (dom96) [inline src/nimblepkg/publish.nim]: This sounds insanely unreliable.

**Comment by dom96** [comment]
In general, very good and very cool, but would be nice to get oauth support. Would also be safer.

File: src/nimblepkg/version.nim (hunk 1 above)
Comment (Araq) [inline src/nimblepkg/version.nim]: Perhaps not necessary now, but necessary for the upcoming PRs. Also I think it does make sense here too.

File: src/nimblepkg/publish.nim (hunk 6 above)
Comment (Araq) [inline src/nimblepkg/publish.nim]: True but works in practice and github recommends 5 seconds. But on a second though, we can busy wait for 1 minute, checking from time to time 'forkExists'. But I think it's good enough for now. Let people report problems.

File: src/nimblepkg/publish.nim (hunk 5 above)
Comment (Araq) [inline src/nimblepkg/publish.nim]: Nope, we don't want the fork to be in a temporary directory.

File: src/nimblepkg/publish.nim (hunk 4 above)
Comment (Araq) [inline src/nimblepkg/publish.nim]: I don't like the dependency and it's been tested with this code. Testing this cannot be done effectively with unit tests and took as much time as writing it.

File: src/nimblepkg/publish.nim (hunk 2 above)
Comment (Araq) [inline src/nimblepkg/publish.nim]: Yes, but Nimble already requires '-d:ssl'.

Comment (Araq) [inline src/nimblepkg/publish.nim]: Also, I don't have to check. It raises an exception if it fails and the output is a huge JSON full of info I don't care about.

File: src/nimblepkg/publish.nim (hunk 3 above)
Comment (Araq) [inline src/nimblepkg/publish.nim]: Again, tested it this way and it is MUCH better than nothing. Let others complain and improve it.

File: src/nimblepkg/publish.nim (hunk 5 above)
Comment (dom96) [inline src/nimblepkg/publish.nim]: Why?

Comment (Araq) [inline src/nimblepkg/publish.nim]: Because we don't want to 'git clone' all the time.
